File: windows/GlassToKey/TrackpadLayoutPreset.cs
using System;

namespace GlassToKey;

public sealed class TrackpadLayoutPreset
{
    public static TrackpadLayoutPreset Blank { get; } = new(
        name: "Blank",
        displayName: "Blank",
        columns: 0,
        rows: 0,
        columnAnchorsMm: Array.Empty<PointMm>(),
        rightLabels: Array.Empty<string[]>()
    );

    public static TrackpadLayoutPreset SixByThree { get; } = new(
        name: "6x3",
        displayName: "6x3",
        columns: 6,
        rows: 3,
        columnAnchorsMm: new[]
        {
            new PointMm(35.0, 20.9),
            new PointMm(53.0, 19.2),
            new PointMm(71.0, 17.5),
            new PointMm(89.0, 19.2),
            new PointMm(107.0, 22.6),
            new PointMm(125.0, 22.6)
        },
        rightLabels: new[]
        {
            new[] { "Y", "U", "I", "O", "P", "Back" },
            new[] { "H", "J", "K", "L", ";", "Ret" },
            new[] { "N", "M", ",", ".", "/", "Ret" }
        }
    );

    public static TrackpadLayoutPreset SixByFour { get; } = new(
        name: "6x4",
        displayName: "6x4",
        columns: 6,
        rows: 4,
        columnAnchorsMm: new[]
        {
            new PointMm(32.0, 14.0),
            new PointMm(50.0, 14.0),
            new PointMm(68.0, 14.0),
            new PointMm(86.0, 14.0),
            new PointMm(104.0, 14.0),
            new PointMm(122.0, 14.0)
        },
        rightLabels: new[]
        {
            new[] { "6", "7", "8", "9", "0", "Back" },
            new[] { "Y", "U", "I", "O", "P", "]" },
            new[] { "H", "J", "K", "L", ";", "Ret" },
            new[] { "N", "M", ",", ".", "/", "Ret" }
        }
    );

    public static TrackpadLayoutPreset FiveByThree { get; } = new(
        name: "5x3",
        displayName: "5x3",
        columns: 5,
        rows: 3,
        columnAnchorsMm: new[]
        {
            new PointMm(36.0, 19.0),
            new PointMm(56.0, 19.0),
            new PointMm(76.0, 19.0),
            new PointMm(96.0, 19.0),
            new PointMm(116.0, 19.0)
        },
        rightLabels: new[]
        {
            new[] { "Y", "U", "I", "O", "P" },
            new[] { "H", "J", "K", "L", ";" },
            new[] { "N", "M", ",", ".", "/" }
        }
    );

    public static TrackpadLayoutPreset FiveByFour { get; } = new(
        name: "5x4",
        displayName: "5x4",
        columns: 5,
        rows: 4,
        columnAnchorsMm: new[]
        {
            new PointMm(34.0, 12.0),
            new PointMm(54.0, 12.0),
            new PointMm(74.0, 12.0),
            new PointMm(94.0, 12.0),
            new PointMm(114.0, 12.0)
        },
        rightLabels: new[]
        {
            new[] { "5", "6", "7", "8", "9" },
            new[] { "T", "Y", "U", "I", "O" },
            new[] { "G", "H", "J", "K", "L" },
            new[] { "B", "N", "M", ",", "." }
        }
    );

    public static TrackpadLayoutPreset Mobile { get; } = new(
        name: "mobile",
        displayName: "Mobile",
        columns: 10,
        rows: 4,
        columnAnchorsMm: new[]
        {
            new PointMm(8.0, 10.0),
            new PointMm(22.0, 10.0),
            new PointMm(36.0, 10.0),
            new PointMm(50.0, 10.0),
            new PointMm(64.0, 10.0),
            new PointMm(78.0, 10.0),
            new PointMm(92.0, 10.0),
            new PointMm(106.0, 10.0),
            new PointMm(120.0, 10.0),
            new PointMm(134.0, 10.0)
        },
        rightLabels: new[]
        {
            new[] { "Q", "W", "E", "R", "T", "Y", "U", "I", "O", "P" },
            new[] { "A", "S", "D", "F", "G", "H", "J", "K", "L", "Ret" },
            new[] { "Z", "X", "C", "V", "B", "N", "M", ",", ".", "Back" },
            new[] { "Space" }
        },
        blankLeftSide: true,
        useFixedRightStaggeredQwerty: true,
        fixedKeyScale: 0.75
    );

    public static TrackpadLayoutPreset MobileOrthoTwelveByFour { get; } = new(
        name: "mobile-ortho-12x4",
        displayName: "Planck",
        columns: 12,
        rows: 4,
        columnAnchorsMm: new[]
        {
            new PointMm(6.0, 10.0),
            new PointMm(18.0, 10.0),
            new PointMm(30.0, 10.0),
            new PointMm(42.0, 10.0),
            new PointMm(54.0, 10.0),
            new PointMm(66.0, 10.0),
            new PointMm(78.0, 10.0),
            new PointMm(90.0, 10.0),
            new PointMm(102.0, 10.0),
            new PointMm(114.0, 10.0),
            new PointMm(126.0, 10.0),
            new PointMm(138.0, 10.0)
        },
        rightLabels: new[]
        {
            new[] { "1", "2", "3", "4", "5", "6", "7", "8", "9", "0", "-", "Back" },
            new[] { "Q", "W", "E", "R", "T", "Y", "U", "I", "O", "P", "[", "]" },
            new[] { "A", "S", "D", "F", "G", "H", "J", "K", "L", ";", "'", "Ret" },
            new[] { "Z", "X", "C", "V", "B", "N", "M", ",", ".", "/", "Shift", "Space" }
        },
        blankLeftSide: true,
        fixedKeyScale: 0.70,
        allowsColumnSettings: true
    );

    public static TrackpadLayoutPreset[] All { get; } =
    {
        Blank,
        SixByThree,
        SixByFour,
        FiveByThree,
        FiveByFour,
        MobileOrthoTwelveByFour,
        Mobile
    };

    public string Name { get; }
    public string DisplayName { get; }
    public int Columns { get; }
    public int Rows { get; }
    public PointMm[] ColumnAnchorsMm { get; }
    public string[][] RightLabels { get; }
    public bool BlankLeftSide { get; }
    public bool UseFixedRightStaggeredQwerty { get; }
    public double FixedKeyScale { get; }
    public bool AllowsColumnSettings { get; }
    public string[][] LeftLabels => Mirror(RightLabels);

    private TrackpadLayoutPreset(
        string name,
        string displayName,
        int columns,
        int rows,
        PointMm[] columnAnchorsMm,
        string[][] rightLabels,
        bool blankLeftSide = false,
        bool useFixedRightStaggeredQwerty = false,
        double fixedKeyScale = 1.0,
        bool? allowsColumnSettings = null)
    {
        Name = name;
        DisplayName = string.IsNullOrWhiteSpace(displayName) ? name : displayName;
        Columns = columns;
        Rows = rows;
        ColumnAnchorsMm = columnAnchorsMm;
        RightLabels = rightLabels;
        BlankLeftSide = blankLeftSide;
        UseFixedRightStaggeredQwerty = useFixedRightStaggeredQwerty;
        FixedKeyScale = Math.Clamp(fixedKeyScale, 0.25, 2.0);
        AllowsColumnSettings = allowsColumnSettings ?? !useFixedRightStaggeredQwerty;
    }

    public override string ToString()
    {
        return DisplayName;
    }

    public static TrackpadLayoutPreset ResolveByNameOrDefault(string? name)
    {
        if (string.IsNullOrWhiteSpace(name))
        {
            return SixByThree;
        }

        string trimmed = name.Trim();
        if (string.Equals(trimmed, "Mobile QWERTY", StringComparison.OrdinalIgnoreCase) ||
            string.Equals(trimmed, "mobile-qwerty", StringComparison.OrdinalIgnoreCase) ||
            string.Equals(trimmed, "mobile_qwerty", StringComparison.OrdinalIgnoreCase))
        {
            return Mobile;
        }

        for (int i = 0; i < All.Length; i++)
        {
            if (string.Equals(All[i].Name, trimmed, StringComparison.OrdinalIgnoreCase) ||
                string.Equals(All[i].DisplayName, trimmed, StringComparison.OrdinalIgnoreCase))
            {
                return All[i];
            }
        }

        return SixByThree;
    }

    private static string[][] Mirror(string[][] labels)
    {
        string[][] output = new string[labels.Length][];
        for (int r = 0; r < labels.Length; r++)
        {
            string[] row = labels[r];
            string[] mirrored = new string[row.Length];
            for (int c = 0; c < row.Length; c++)
            {
                mirrored[c] = row[row.Length - 1 - c];
            }
            output[r] = mirrored;
        }
        return output;
    }
}

public readonly record struct PointMm(double X, double Y);

File: mac/GlassToKey/GlassToKey/TrackpadLayoutPreset.swift
import CoreGraphics

enum TrackpadLayoutPreset: String, CaseIterable, Identifiable {
    case blank = "Blank"
    case sixByThree = "6x3"
    case sixByFour = "6x4"
    case fiveByThree = "5x3"
    case fiveByFour = "5x4"
    case mobileOrtho12x4 = "mobile-ortho-12x4"
    case mobile = "mobile"

    static let allCases: [TrackpadLayoutPreset] = [
        .blank,
        .sixByThree,
        .sixByFour,
        .fiveByThree,
        .fiveByFour,
        .mobileOrtho12x4,
        .mobile
    ]

    static let selectableCases: [TrackpadLayoutPreset] = [
        .blank,
        .sixByThree,
        .sixByFour,
        .fiveByThree,
        .fiveByFour,
        .mobileOrtho12x4
    ]

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .blank:
            return rawValue
        case .mobileOrtho12x4:
            return "Planck"
        case .mobile:
            return "Mobile"
        default:
            return rawValue
        }
    }

    var columns: Int {
        switch self {
        case .sixByThree, .sixByFour:
            return 6
        case .fiveByThree, .fiveByFour:
            return 5
        case .mobileOrtho12x4:
            return 12
        case .mobile:
            return 10
        case .blank:
            return 0
        }
    }

    var rows: Int {
        switch self {
        case .sixByThree, .fiveByThree:
            return 3
        case .sixByFour, .fiveByFour, .mobileOrtho12x4, .mobile:
            return 4
        case .blank:
            return 0
        }
    }

    var hasGrid: Bool {
        columns > 0 && rows > 0
    }

    var columnAnchors: [CGPoint] {
        switch self {
        case .sixByThree, .sixByFour:
            return Self.columnAnchors6
        case .fiveByThree:
            return Self.columnAnchors5FromSixTakeFirst
        case .fiveByFour:
            return Self.columnAnchors5FromSixTakeFirst
        case .mobileOrtho12x4:
            return Self.columnAnchors12
        case .mobile:
            return Self.columnAnchors10
        case .blank:
            return []
        }
    }

    var rightLabels: [[String]] {
        switch self {
        case .sixByThree:
            return Self.rightLabels6x3
        case .sixByFour:
            return Self.rightLabels6x4
        case .fiveByThree:
            return Self.rightLabels6x3.map { Array($0.prefix(self.columns)) }
        case .fiveByFour:
            return Self.rightLabels6x4.map { Array($0.prefix(self.columns)) }
        case .mobileOrtho12x4:
            return Self.rightLabels12x4
        case .mobile:
            return Self.rightLabelsMobile
        case .blank:
            return []
        }
    }

    var leftLabels: [[String]] {
        switch self {
        case .mobileOrtho12x4, .mobile:
            return []
        default:
            return Self.mirrored(rightLabels)
        }
    }

    var blankLeftSide: Bool {
        switch self {
        case .mobileOrtho12x4, .mobile:
            return true
        default:
            return false
        }
    }

    var allowsColumnSettings: Bool {
        switch self {
        case .mobile:
            return false
        default:
            return hasGrid
        }
    }

    static func resolveByName(_ name: String?) -> TrackpadLayoutPreset? {
        guard let trimmed = name?.trimmingCharacters(in: .whitespacesAndNewlines),
              !trimmed.isEmpty else {
            return nil
        }

        switch trimmed.lowercased() {
        case "blank", "none":
            return .blank
        case "mobile-ortho-12x4", "mobile ortho 12x4":
            return .mobileOrtho12x4
        default:
            return allCases.first { $0.rawValue.caseInsensitiveCompare(trimmed) == .orderedSame }
        }
    }

    static func resolveByNameOrDefault(_ name: String?) -> TrackpadLayoutPreset {
        resolveByName(name) ?? .sixByThree
    }

    var allowHoldBindings: Bool {
        true
    }

    private static func mirrored(_ labels: [[String]]) -> [[String]] {
        labels.map { Array($0.reversed()) }
    }

    private static let columnAnchors6: [CGPoint] = [
        CGPoint(x: 35.0, y: 20.9),
        CGPoint(x: 53.0, y: 19.2),
        CGPoint(x: 71.0, y: 17.5),
        CGPoint(x: 89.0, y: 19.2),
        CGPoint(x: 107.0, y: 22.6),
        CGPoint(x: 125.0, y: 22.6)
    ]

    private static let columnAnchors5FromSixDropFirst: [CGPoint] = Array(columnAnchors6.suffix(5))
    private static let columnAnchors5FromSixTakeFirst: [CGPoint] = Array(columnAnchors6.prefix(5))

    private static let columnAnchors12: [CGPoint] = [
        CGPoint(x: 6.0, y: 10.0),
        CGPoint(x: 18.0, y: 10.0),
        CGPoint(x: 30.0, y: 10.0),
        CGPoint(x: 42.0, y: 10.0),
        CGPoint(x: 54.0, y: 10.0),
        CGPoint(x: 66.0, y: 10.0),
        CGPoint(x: 78.0, y: 10.0),
        CGPoint(x: 90.0, y: 10.0),
        CGPoint(x: 102.0, y: 10.0),
        CGPoint(x: 114.0, y: 10.0),
        CGPoint(x: 126.0, y: 10.0),
        CGPoint(x: 138.0, y: 10.0)
    ]

    private static let columnAnchors10: [CGPoint] = [
        CGPoint(x: 8.0, y: 10.0),
        CGPoint(x: 22.0, y: 10.0),
        CGPoint(x: 36.0, y: 10.0),
        CGPoint(x: 50.0, y: 10.0),
        CGPoint(x: 64.0, y: 10.0),
        CGPoint(x: 78.0, y: 10.0),
        CGPoint(x: 92.0, y: 10.0),
        CGPoint(x: 106.0, y: 10.0),
        CGPoint(x: 120.0, y: 10.0),
        CGPoint(x: 134.0, y: 10.0)
    ]

    private static let rightLabels6x3: [[String]] = [
        ["Y", "U", "I", "O", "P", "Back"],
        ["H", "J", "K", "L", ";", "Ret"],
        ["N", "M", ",", ".", "/", "Ret"]
    ]

    private static let rightLabels6x4: [[String]] = [
        ["6", "7", "8", "9", "0", "Back"],
        ["Y", "U", "I", "O", "P", "]"],
        ["H", "J", "K", "L", ";", "Ret"],
        ["N", "M", ",", ".", "/", "Ret"]
    ]

    private static let rightLabels12x4: [[String]] = [
        ["1", "2", "3", "4", "5", "6", "7", "8", "9", "0", "-", "Back"],
        ["Q", "W", "E", "R", "T", "Y", "U", "I", "O", "P", "[", "]"],
        ["A", "S", "D", "F", "G", "H", "J", "K", "L", ";", "'", "Ret"],
        ["Z", "X", "C", "V", "B", "N", "M", ",", ".", "/", "Shift", "Space"]
    ]

    private static let rightLabelsMobile: [[String]] = [
        ["Q", "W", "E", "R", "T", "Y", "U", "I", "O", "P"],
        ["A", "S", "D", "F", "G", "H", "J", "K", "L", "Ret"],
        ["Z", "X", "C", "V", "B", "N", "M", ",", ".", "Back"],
        ["Space"]
    ]
}
